File: library/src/main/java/com/grosner/dbflow/sql/StringQuery.java
package com.grosner.dbflow.sql;

import android.database.Cursor;

import com.grosner.dbflow.config.FlowManager;
import com.grosner.dbflow.structure.Model;

import java.util.List;

/**
 * Author: andrewgrosner
 * Description: Provides a very basic query mechanism for strings. It runs a modification query and will only run
 * {@link android.database.sqlite.SQLiteDatabase#rawQuery(String, String[])}.
 */
public class StringQuery<ModelClass extends Model> implements Query, Queriable<ModelClass> {

    /**
     * The full SQLite query to use
     */
    private final String mQuery;

    private final Class<ModelClass> mTable;

    /**
     * Creates an instance of this class
     *
     * @param table The table to use
     * @param sql   The sql statement to query the DB with. Does not work with {@link com.grosner.dbflow.sql.language.Delete},
     *              this must be done with {@link android.database.sqlite.SQLiteDatabase#execSQL(String)}
     */
    public StringQuery(Class<ModelClass> table, String sql) {
        mQuery = sql;
        mTable = table;
    }

    @Override
    public String getQuery() {
        return mQuery;
    }

    @Override
    public Cursor query() {
        return FlowManager.getDatabaseForTable(mTable).getWritableDatabase().rawQuery(mQuery, null);
    }

    @Override
    public List<ModelClass> queryList() {
        return SqlUtils.queryList(mTable, mQuery);
    }

    @Override
    public ModelClass querySingle() {
        return SqlUtils.querySingle(mTable, mQuery);
    }
}

File: lib/src/main/kotlin/com/dbflow5/query/StringQuery.kt
package com.dbflow5.query

import android.database.sqlite.SQLiteDatabase
import com.dbflow5.adapter.RetrievalAdapter
import com.dbflow5.config.FlowLog
import com.dbflow5.database.DatabaseStatement
import com.dbflow5.database.DatabaseStatementWrapper
import com.dbflow5.database.DatabaseWrapper
import com.dbflow5.database.FlowCursor
import com.dbflow5.sql.Query
import com.dbflow5.structure.ChangeAction
import kotlinx.coroutines.runBlocking

/**
 * Description: Provides a very basic query mechanism for strings. Allows you to easily perform custom SQL cursor string
 * code where this library does not provide. It only runs a
 * [SQLiteDatabase.rawQuery].
 */
class StringQuery<T : Any>
/**
 * Creates an instance of this class
 *
 * @param table The table to use
 * @param query   The sql statement to query the DB with. Does not work with [Delete],
 * this must be done with [DatabaseWrapper.execSQL]
 */
    (
    adapter: RetrievalAdapter<T>,
    override val query: String
) : BaseModelQueriable<T>(adapter), Query, ModelQueriable<T> {
    private var args: Array<String>? = null

    override// we don't explicitly know the change, but something changed.
    val primaryAction: ChangeAction
        get() = ChangeAction.CHANGE

    override fun cursor(databaseWrapper: DatabaseWrapper): FlowCursor =
        databaseWrapper.rawQuery(query, args)

    override fun queryList(databaseWrapper: DatabaseWrapper): List<T> {
        FlowLog.log(FlowLog.Level.V, "Executing query: $query")
        return runBlocking { adapter.loadList(cursor(databaseWrapper), databaseWrapper) }
    }

    override fun querySingle(databaseWrapper: DatabaseWrapper): T? {
        FlowLog.log(FlowLog.Level.V, "Executing query: $query")
        return runBlocking { adapter.loadSingle(cursor(databaseWrapper), databaseWrapper) }
    }

    override fun <QueryClass : Any> queryCustomList(
        retrievalAdapter: RetrievalAdapter<QueryClass>,
        databaseWrapper: DatabaseWrapper
    )
        : List<QueryClass> {
        val query = query
        FlowLog.log(FlowLog.Level.V, "Executing query: $query")
        return runBlocking { retrievalAdapter.loadList(cursor(databaseWrapper), databaseWrapper) }
    }

    override fun <QueryClass : Any> queryCustomSingle(
        retrievalAdapter: RetrievalAdapter<QueryClass>,
        databaseWrapper: DatabaseWrapper
    )
        : QueryClass? {
        val query = query
        FlowLog.log(FlowLog.Level.V, "Executing query: $query")
        return runBlocking { retrievalAdapter.loadSingle(cursor(databaseWrapper), databaseWrapper) }
    }

    override fun compileStatement(databaseWrapper: DatabaseWrapper): DatabaseStatement {
        val query = query
        FlowLog.log(
            FlowLog.Level.V,
            "Compiling Query Into Statement: query = $query , args = $args"
        )
        return DatabaseStatementWrapper(
            databaseWrapper.compileStatement(query, args), this,
            databaseWrapper
        )
    }

    /**
     * Set selection arguments to execute on this raw query.
     */
    fun setSelectionArgs(args: Array<String>) = apply {
        this.args = args
    }
}
